content/blogposts/2023/copilot-vs-cody.md

```diff
@@ -0,0 +1,229 @@
+
+Both Cody and Copilot were able to understand our query and provided a working solution. There is one subtle difference that gives Cody a major advantage though. We are creating a Laravel Action, which is written in PHP. Cody has given us the code as a PHP function. Copilot on the other hand gave us a working implementation in JavaScript.
+
+This is another perfect example of why context is so important. Cody inferred that since we have a `.php` file open, we probably want our implementation in PHP, whereas Copilot lacking the context gave us a general solution in a different language. If we were to insert the two functions into our file, the Cody `millisecondsToTimecode` function would work, but the Copilot `msToTimecode` would throw all sorts of errors. Cody wins here as well, but we'll give Copilot half a point because it did give us a working solution.
+
+Cody - 4
+
+Copilot - 0.5
+
+## Scenario: Editing code in our codebase
+
+We'll use the `millisecondsToTimecode` function that Cody gave us and continue adding functionality to our application. The timecode conversion function that both Cody and Copilot returned gives us a string in the format of "00:00:00" and while this is valid, we should also add the number of frames between a second. This will ensure that we trim the video at max precision. So the full format becomes hour:minute:second:frame.
+
+To get the frame count we'll need to know the frame rate at which the video was encoded as well as what the remainder of milliseconds if any we have for our trim operation. Let's ask Cody and Copilot to update the existing function to accommodate our needs.
+
+![Using code AI to edit functionality](https://storage.googleapis.com/sourcegraph-assets/blog/github-copilot-vs-cody-context-driven-development/edit-function-with-new-functionality.png)
+
+For Cody, with my `TrimVideoAction.php` file open, I was able to say **"add the frame number to the timecode"** and Cody had enough context to understand that I was talking about the `millisecondsToTimecode()` function and it gave me an updated function that also added the number of left-over frames if any.
+
+For Copilot, I tried the same query, but Copilot could not understand that I wanted to update the `millisecondsToTimecode()` function, so eventually I had to ask the question and copy and paste the function into the chat window to get a working result. Copilot was able to understand my prompt and returned an updated `millisecondsToTimecode()` function that also returned the frame count. The one thing that I think Copilot did better was it made the frame rate a parameter. Videos typically have varying frame rates typically ranging from 24fps to 30fps, and the frame rate is something we store in the `Video` model as a value because each video is different, so this was a nice touch.
```

Comment:
```suggestion
For Copilot, I tried the same query, but Copilot could not understand that I wanted to update the `millisecondsToTimecode()` function, so eventually I had to ask the question and copy and paste the function into the chat window to get a working result. Copilot was able to understand my prompt and returned an updated `millisecondsToTimecode()` function that also returned the frame count. The one thing that I think Copilot did better was it made the frame rate a parameter. Videos typically have varying frame rates ranging from 24fps to 30fps, and the frame rate is something we store in the `Video` model as a value because each video is different, so this was a nice touch.
```

Reply:
I will keep typically as a lot of videos are 24-30fps, but videos can be any fps like 60, 70, 100+.